*** a/cups/ipp.h	2014-04-09 20:12:27.000000000 +0400
--- b/cups/ipp.h	2016-06-01 19:15:50.578545836 +0300
***************
*** 60,66 ****
  #  define IPP_MAX_LANGUAGE	64	/* Maximum length of naturalLanguage values w/nul */
  #  define IPP_MAX_LENGTH	32767	/* Maximum size of any single value */
  #  define IPP_MAX_MIMETYPE	256	/* Maximum length of mimeMediaType values w/nul */
! #  define IPP_MAX_NAME		256	/* Maximum length of common name values w/nul */
  #  define IPP_MAX_OCTETSTRING	1023	/* Maximum length of octetString values w/o nul */
  #  define IPP_MAX_TEXT		1024	/* Maximum length of text values w/nul */
  #  define IPP_MAX_URI		1024	/* Maximum length of uri values w/nul */
--- 60,66 ----
  #  define IPP_MAX_LANGUAGE	64	/* Maximum length of naturalLanguage values w/nul */
  #  define IPP_MAX_LENGTH	32767	/* Maximum size of any single value */
  #  define IPP_MAX_MIMETYPE	256	/* Maximum length of mimeMediaType values w/nul */
! #  define IPP_MAX_NAME		1024	/* Maximum length of common name values w/nul */
  #  define IPP_MAX_OCTETSTRING	1023	/* Maximum length of octetString values w/o nul */
  #  define IPP_MAX_TEXT		1024	/* Maximum length of text values w/nul */
  #  define IPP_MAX_URI		1024	/* Maximum length of uri values w/nul */
